### ISSUE TITLE
Tensor type annotations

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn_common#97]
`out_shape` or `assert_shape` is about checking the complete shape including all dims. The user or developer might use this as a kind of verification.

The API of any module or function might (should) however be flexible to any number of dimensions. Basically, for some input tensor, there are maybe one or more dims which are relevant (`in_dim`, `in_spatial_dims`, `axis`, or whatever) and there could also be any other number of dims which will then be treated just as the batch dim.

This is maybe a separate aspect. The definition of module/function API is maybe not covered well yet. The usual syntax is via type annotations. There this new Python PEP is relevant, although we would want to differ in a few aspects:

- Type annotations are usually only checked statically by external software (MyPy or PyCharm). We might want to check it always, also at runtime.
- We would not define a shape tuple but a set of dims for input and output.
- This set of dims would not be complete. There could always be other dims like batch dim and others.
- In the output shape, the same.
- This PEP is only supported since Python 3.10 but we want to support Python >=3.6 or so.

One question is whether we should somehow be more explicit when there are multiple inputs, whether the output would include all dims from all inputs, or sth else. E.g. maybe it could look like:

```
func(a: Tensor[{A, B, *more_a}], b: Tensor[{A, B, *more_b}])
  -> Tensor[{A, B, *more_a.union(more_b)}]
```

Or simpler:

```
func(a: Tensor[{A, B, Any}], b: Tensor[{A, B, Any}])
  -> Tensor[{A, B, Any}]
```

With the assumption that the output would always be the union of the inputs.

Or more specific:

```
func(a: Tensor[{A, B, Any}], b: Tensor[{A, B, Any}])
  -> Tensor[{A, B, Any(a, b)}]
```

Or even more simpler:

```
func(a: Tensor[{A, B}], b: Tensor[{A, B}])
  -> Tensor[{A, B}]
```

With the assumption that there always can be additional dims.

Further, such type annotation might include other information like dtype.

_Originally posted by @albertz in https://github.com/rwth-i6/returnn_common/issues/17#issuecomment-1025125463_

**Comment by albertz** [comment]
Often some dimension in the inputs is actually also an argument to the function. It's not so clear how this should look like:

```
dot(a: Tensor[{A, B}], b: Tensor[{B, C}], reduce: B)
  -> Tensor[{A, C}]
```
?

**Comment by albertz** [comment]
I tend to postpone such a feature for now, and not make it part of the first release. There are too many open questions.

**Comment by JackTemaki** [comment]
> I tend to postpone such a feature for now, and not make it part of the first release.

I agree with this, `returnn_common` is already very complex (maybe too complex) in its current state, and we should first aim for a version we can actually test setups with as soons as possible.
